src/components/search-bar/search-bar.tsx

```diff
@@ -1,52 +1,130 @@
-import React, { useState, KeyboardEvent, ChangeEvent } from 'react';
+import React, { useState, KeyboardEvent, ChangeEvent, useMemo, useRef } from 'react';
 import { ReactComponent as CancelIcon } from '../../assets/svg/cancel-icon.svg';
 import { ReactComponent as SearchIcon } from '../../assets/svg/search-icon.svg';
+import { debounce, noop } from '../../helpers/func';
+import { includesIgnoreCase } from '../../helpers/string';
+import { useOutsideClick } from '../../hooks/use-outside-click';
+import { DropDownOption } from '../drop-down/drop-down';
 import './search-bar.scss';
 
 export interface SearchBarProps {
   initialText?: string;
   placeholder?: string;
   disabled?: boolean;
+  dropDownData?: DropDownOption[];
+  dropDownIgnoreCase?: boolean;
+  debounceMs?: number;
   onChange?: (value: string) => void;
   onEnterPressed?: (value: string) => void;
+  onDebouncedChange?: (value: string) => void;
+  onDropdownClick?: (option: DropDownOption) => void;
 }
 
 export const SearchBar = ({
-  initialText,
+  initialText = '',
   placeholder,
   disabled,
+  dropDownData,
+  dropDownIgnoreCase = true,
+  debounceMs = 250,
   onChange,
   onEnterPressed,
+  onDebouncedChange,
+  onDropdownClick,
 }: SearchBarProps) => {
   const [value, setValue] = useState(initialText);
+  const [hasClickedSinceLastChange, setHasClickedSinceLastChange] = useState(false);
+
+  const debouncedChange = useMemo(
+    () => debounce(onDebouncedChange ?? noop, debounceMs),
+    [onDebouncedChange, debounceMs]
+  );
+
+  // hide dropdown after clicking outside the search bar wrapper div
+  const wrapperDivRef = useRef(null);
+  useOutsideClick(wrapperDivRef, () => setHasClickedSinceLastChange(true));
```

Comment:
This useOutsideClick is so nice. I think I will add that to the flashcard component if we decide to do an onBlur thing

Reply:
I didn't realize you made that yourself; that's so dope